File: LC/1981/1981.java
/*
[1,2,3],
[4,5,6],
[7,8,9]
*/
class Solution {
    public int minimizeTheDifference(int[][] mat, int target) {
        if (mat.length == 0 || mat[0].length == 0) return target;
        Integer[][] dp = new Integer[mat.length][4900];
        return minimize(mat, 0, 0, target, dp);
    }
    
    private int minimize(int[][] mat, int row, int sum, int target, Integer[][] dp) {
        if (row == mat.length) return Math.abs(target - sum);
        if (dp[row][sum] != null) return dp[row][sum];
        
        int minVal = Integer.MAX_VALUE;
        for (int col = 0; col < mat[0].length; col++) {
            int val = mat[row][col];
            minVal = Math.min(minVal, minimize(mat, row+1, sum+val, target, dp));
        }
        dp[row][sum] = minVal;
        return minVal;
    }
}

File: LC/1981/1981.ts
/*
[1,2,3]
[4,5,6]
[7,8,9]

[1]
[5,6,7]
[12,13,14]

[1,0,1]
                 1          2          3
[0,...,4900]
[]
                 
*/
function minimizeTheDifference(mat: number[][], target: number): number {
    const dp = new Array(mat.length).fill(null).map((row) => new Array(4900).fill(null));
    return dfs(mat, 0, target, 0, dp);
};

function dfs(mat: number[][], row: number, target: number, sum: number, dp: number[][]): number {
    if (row >= mat.length) return Math.abs(target-sum);
    
    if (dp[row][sum] !== null) return dp[row][sum];
    
    let smallestDiff = Number.MAX_SAFE_INTEGER;
    for (let col = 0; col < mat[0].length; col++) {
        smallestDiff = Math.min(smallestDiff, dfs(mat, row+1, target, sum+mat[row][col], dp));
    }
    dp[row][sum] = smallestDiff;
    return smallestDiff;
};
